pkg/netceptor/conn.go

```diff
@@ -412,27 +421,49 @@ func (c *Conn) CancelRead() {
 }
 
 // Write writes data to the connection.
-func (c *Conn) Write(b []byte) (n int, err error) {
-	return c.qs.Write(b)
+func (c *Conn) Write(b []byte) (int, error) {
+	n, err := c.qs.Write(b)
+	aerr, ok := err.(*quic.ApplicationError)
+	if ok && aerr.ErrorMessage == "normal close" {
+		err = net.ErrClosed
+	}
+
+	return n, err
 }
 
-// Close closes the writer side of the connection.
-func (c *Conn) Close() error {
+// CloseWrite closes the writer side of the connection.
+func (c *Conn) CloseWrite() error {
 	c.doneOnce.Do(func() {
 		close(c.doneChan)
 	})
 
 	return c.qs.Close()
 }
 
-func (c *Conn) CloseConnection() error {
-	c.pc.cancel()
+// Close closes the connection.
+func (c *Conn) Close() error {
+	if !c.fromListener {
```

Comment:
Can you explain why this packet conn only can be cancelled this way please?

I am struggling to understand when `fromListener` will be set to `false`

Reply:
If I remember right, fromListener is set based on whether a Conn object was created by the user opening it, or was created by a listener receiving a connection.  If the Conn is standalone, then when we close it, we need to close the underlying PacketConn.  But if the Conn came from a listener, the PacketConn is shared, and should only be closed when the listener is closed.